libs/gl-client-py/src/credentials.rs

```diff
@@ -1,147 +1,145 @@
+        }
     }
+}
 
-    pub fn build(&self) -> Result<Credentials> {
-        let inner = self.inner.clone().build()?;
-        Ok(Credentials {
-            typ: CredentialType::Device,
-            inner,
-        })
+impl<T, R> RuneProvider for UnifiedCredentials<T, R>
+where
+    T: TlsConfigProvider,
+    R: TlsConfigProvider + RuneProvider + Clone,
+{
+    fn rune(&self) -> String {
+        match self {
+            UnifiedCredentials::Nobody(_) => panic!(
+                "can not provide rune from nobody credentials! something really bad happended."
+            ),
+            UnifiedCredentials::Device(d) => d.rune(),
+        }
     }
 }
 
 #[pyclass]
-pub struct NobodyBuilder {
-    inner: credentials::Builder<credentials::Nobody>,
+#[derive(Clone)]
+pub struct Credentials {
+    pub inner: PyCredentials,
 }
 
 #[pymethods]
-impl NobodyBuilder {
+impl Credentials {
     #[new]
     pub fn new() -> Self {
-        Self {
-            inner: credentials::Builder::as_nobody(),
-        }
-    }
-
-    pub fn with_default(&self) -> Result<Self> {
-        let inner = self.inner.clone().with_default()?;
-        Ok(Self { inner })
-    }
-
-    pub fn with_identity(&self, cert: &[u8], key: &[u8]) -> Result<Self> {
-        let inner = self.inner.clone().with_identity(cert, key);
-        Ok(Self { inner })
-    }
-
-    pub fn with_ca(&self, ca: &[u8]) -> Result<Self> {
-        let inner = self.inner.clone().with_ca(ca);
-        Ok(Self { inner })
+        let inner = UnifiedCredentials::Nobody(gl_client::credentials::Nobody::default());
+        Self { inner }
     }
 
-    pub fn build(&self) -> Result<Credentials> {
-        let inner = self.inner.clone().build()?;
-        Ok(Credentials {
-            typ: CredentialType::Nobody,
-            inner,
-        })
+    #[staticmethod]
+    pub fn nobody_with(cert: &[u8], key: &[u8], ca: &[u8]) -> Self {
+        let inner = UnifiedCredentials::Nobody(gl_client::credentials::Nobody::with(cert, key, ca));
+        Self { inner }
     }
-}
-
-#[pyclass]
-#[derive(Clone)]
-pub struct Credentials {
-    pub typ: CredentialType,
-    pub inner: credentials::Credentials,
-}
 
-#[pymethods]
-impl Credentials {
     #[staticmethod]
-    pub fn as_device() -> DeviceBuilder {
-        DeviceBuilder::new()
+    pub fn from_path(path: &str) -> Self {
+        let inner = UnifiedCredentials::Device(gl_client::credentials::Device::from_path(path));
+        Self { inner }
     }
 
     #[staticmethod]
-    pub fn as_nobody() -> NobodyBuilder {
-        NobodyBuilder::new()
+    pub fn from_bytes(data: &[u8]) -> Self {
+        let inner = UnifiedCredentials::Device(gl_client::credentials::Device::from_bytes(data));
+        Self { inner }
     }
 
-    pub fn tls_config(&self) -> Result<TlsConfig> {
-        let inner = self.inner.tls_config()?;
-        Ok(TlsConfig { inner })
+    pub fn upgrade(&self, scheduler: &Scheduler, signer: &Signer) -> Result<Credentials> {
```

Comment:
I'm a bit confused here. 

What is the process of upgrading `credentials` doing?
Why would I want to `upgrade` my credentials?

Reply:
This method exists to give us some compatibility on future changes. If the binary format of the credentials change, or if we add new fields, we have a function that can take care of reflecting these changes. Right now what it does is to call `recover` on the scheduler service to get a full set of credentials containing a rune if we don't have a rune yet.